blockchain/src/blockchain/push.rs

```diff
@@ -261,34 +276,63 @@ impl Blockchain {
         chunk_result
     }
 
-    fn push_wrapperfn(
+    fn push_wrapperfn<F: PostValidationHook>(
         this: RwLockUpgradableReadGuard<Self>,
         block: Block,
         trust: bool,
         diff: Option<TrieDiff>,
         chunks: Vec<TrieChunkWithStart>,
+        post_validation_hook: &F,
     ) -> Result<(PushResult, Result<ChunksPushResult, ChunksPushError>), PushError> {
         #[cfg(not(feature = "metrics"))]
         {
-            Self::do_push(this, block, trust, diff, chunks)
+            let res = Self::do_push(
+                this,
+                block.clone(),
+                trust,
+                diff,
+                chunks,
+                post_validation_hook,
+            );
+            match res {
+                // Extended and Rebranched are already handled pre-commit.
+                Ok((PushResult::Extended | PushResult::Rebranched, _)) => {}
+                Ok((ref res, _)) => post_validation_hook.post_validation(&block, &Ok(res.clone())),
+                Err(ref res) => post_validation_hook.post_validation(&block, &Err(res.clone())),
```

Comment:
Compilation error when the metrics feature is disabled.

```suggestion
                Ok((ref res, _)) => post_validation_hook.post_validation(&block, Ok(res)),
                Err(ref res) => post_validation_hook.post_validation(&block, Err(res)),
```